internal/offline/zip.go

```diff
@@ -0,0 +1,249 @@
+package offline
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"hash/crc32"
+	"io"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/google/osv-scanner/pkg/lockfile"
+	"github.com/google/osv-scanner/pkg/models"
+	"github.com/google/osv-scanner/pkg/osv"
+)
+
+type ZipDB struct {
+	Name            string
+	ArchiveURL      string
+	Offline         bool
+	StoredAt        string
+	vulnerabilities []models.Vulnerability
+}
+
+var ErrOfflineDatabaseNotFound = errors.New("no offline version of the OSV database is available")
+
+func fetchRemoteArchiveCRC32CHash(url string) (uint32, error) {
+	req, err := http.NewRequestWithContext(context.Background(), http.MethodHead, url, nil)
+
+	if err != nil {
+		return 0, err
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return 0, err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return 0, fmt.Errorf("db host returned %s", resp.Status)
+	}
+
+	for _, value := range resp.Header.Values("x-goog-hash") {
+		if strings.HasPrefix(value, "crc32c=") {
+			value = strings.TrimPrefix(value, "crc32c=")
+			out, err := base64.StdEncoding.DecodeString(value)
+
+			if err != nil {
+				return 0, fmt.Errorf("could not decode crc32c= checksum: %w", err)
+			}
+
+			return binary.BigEndian.Uint32(out), nil
+		}
+	}
+
+	return 0, fmt.Errorf("could not find crc32c= checksum")
+}
+
+func fetchLocalArchiveCRC32CHash(data []byte) uint32 {
+	return crc32.Checksum(data, crc32.MakeTable(crc32.Castagnoli))
+}
+
+func (db *ZipDB) fetchZip() ([]byte, error) {
+	cache, err := os.ReadFile(db.StoredAt)
+
+	if db.Offline {
+		if err != nil {
+			return nil, ErrOfflineDatabaseNotFound
+		}
+
+		return cache, nil
+	}
+
+	if err == nil {
+		remoteHash, err := fetchRemoteArchiveCRC32CHash(db.ArchiveURL)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if fetchLocalArchiveCRC32CHash(cache) == remoteHash {
+			return cache, nil
+		}
+	}
+
+	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, db.ArchiveURL, nil)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve OSV database archive: %w", err)
+	}
+
+	if osv.RequestUserAgent != "" {
+		req.Header.Set("User-Agent", osv.RequestUserAgent)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve OSV database archive: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("db host returned %s", resp.Status)
+	}
+
+	defer resp.Body.Close()
```

Comment:
Should this go before line 110? There can be a body opened even if the status vcode isn't 200 right?